File: android/src/main/java/com/margelo/nitro/ahap/Ahap.kt
package com.margelo.nitro.ahap
  
import com.facebook.proguard.annotations.DoNotStrip

@DoNotStrip
class Ahap : HybridAhapSpec() {
  override val memorySize: Long
    get() = 0L

  private var hapticsEnabled: Boolean = true

  override fun startHaptic(events: Array<HapticEvent>, curves: Array<HapticCurve>) {
    // TODO: Android haptics implementation not yet supported
  }

  override fun stopAllHaptics() {
    // TODO: Android haptics implementation not yet supported
  }

  override fun initializeEngine() {
    // TODO: Android haptics implementation not yet supported
  }

  override fun destroyEngine() {
    // TODO: Android haptics implementation not yet supported
  }

  override fun createContinuousPlayer(playerId: String, initialIntensity: Double, initialSharpness: Double) {
    // TODO: Android haptics implementation not yet supported
  }

  override fun startContinuousPlayer(playerId: String) {
    // TODO: Android haptics implementation not yet supported
  }

  override fun updateContinuousPlayer(playerId: String, intensityControl: Double, sharpnessControl: Double) {
    // TODO: Android haptics implementation not yet supported
  }

  override fun stopContinuousPlayer(playerId: String) {
    // TODO: Android haptics implementation not yet supported
  }
  
  override fun destroyContinuousPlayer(playerId: String) {
    // TODO: Android haptics implementation not yet supported
  }

  // MARK: - Global Haptics Enable/Disable

  override fun setHapticsEnabled(enabled: Boolean) {
    hapticsEnabled = enabled
  }

  override fun getHapticsEnabled(): Boolean {
    return hapticsEnabled
  }

  // MARK: - System Haptics (Predefined OS-level feedback)

  override fun triggerImpact(style: HapticImpactStyle) {
    // TODO: Android haptics implementation not yet supported
  }

  override fun triggerNotification(type: HapticNotificationType) {
    // TODO: Android haptics implementation not yet supported
  }

  override fun triggerSelection() {
    // TODO: Android haptics implementation not yet supported
  }
}

File: ios/Ahap.swift
import CoreHaptics

struct AnyHapticAnimationState: HapticAnimationState {
    var hapticEvents: [CHHapticEvent]
    var hapticCurves: [CHHapticParameterCurve]

    init(hapticEvents: [CHHapticEvent], hapticCurves: [CHHapticParameterCurve]) {
        self.hapticEvents = hapticEvents
        self.hapticCurves = hapticCurves
    }
}


class Ahap: HybridAhapSpec {
  func stopAllHaptics() throws {
    haptics.stopAllHapticPlayers()
  }

  func initializeEngine() throws {
    haptics.createAndStartHapticEngine()
  }

  func destroyEngine() throws {
    haptics.clearHapticPlayers()
  }

  // MARK: - Continuous Player Methods

  func createContinuousPlayer(initialIntensity: Double, initialSharpness: Double) throws {
    haptics.createContinuousPlayer(initialIntensity: Float(initialIntensity), initialSharpness: Float(initialSharpness))
  }

  func startContinuousPlayer() throws {
    haptics.startContinuousPlayer()
  }

  func updateContinuousPlayer(intensityControl: Double, sharpnessControl: Double) throws {
    haptics.updateContinuousPlayer(intensityControl: Float(intensityControl), sharpnessControl: Float(sharpnessControl))
  }

  func stopContinuousPlayer() throws {
    haptics.stopContinuousPlayer()
  }

  
  func startHaptic(events: [HapticEvent], curves: [HapticCurve]) throws {
    let hapticEvents = events.map { event -> CHHapticEvent in
    let eventType: CHHapticEvent.EventType = (event.type == .continuous) ? .hapticContinuous : .hapticTransient
    let parameters = event.parameters.map { parameter -> CHHapticEventParameter in
        let parameterID: CHHapticEvent.ParameterID = (parameter.type == .intensity) ? .hapticIntensity : .hapticSharpness
        return CHHapticEventParameter(parameterID: parameterID, value: Float(parameter.value))
    }
    return CHHapticEvent(
    eventType: eventType,
    parameters: parameters,
    relativeTime: event.relativeTime,
    duration: event.type == .continuous ? event.duration ?? 1.0 : 0
    )
}

let hapticCurves = curves.map { curve -> CHHapticParameterCurve in
    let parameterID: CHHapticDynamicParameter.ID = (curve.type == .intensity) ? .hapticIntensityControl : .hapticSharpnessControl
    let controlPoints = curve.controlPoints.map { controlPoint -> CHHapticParameterCurve.ControlPoint in
        return CHHapticParameterCurve.ControlPoint(relativeTime: controlPoint.relativeTime, value: Float(controlPoint.value))
    }
    return CHHapticParameterCurve(
        parameterID: parameterID,
        controlPoints: controlPoints,
        relativeTime: curve.relativeTime
    )
}

let state = AnyHapticAnimationState(hapticEvents: hapticEvents, hapticCurves: hapticCurves)
haptics.createHapticPlayers(for: [state])
haptics.startHapticPlayer(for: state)
  }
}
